src/qt/clientmodel.cpp

```diff
@@ -105,6 +106,14 @@ int64_t ClientModel::getHeaderTipTime() const
     return cachedBestHeaderTime;
 }
 
+int ClientModel::getNumBlocks() const
+{
+    if (m_cached_num_blocks == -1) {
+        m_cached_num_blocks = m_node.getNumBlocks();
```

Comment:
Maybe just initialize `m_cached_num_blocks` after `subscribeToCoreSignals()` and drop the `-1` case?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/17905#discussion_r384785271

> Maybe just initialize `m_cached_num_blocks` after `subscribeToCoreSignals()` and drop the `-1` case?

It seems a little more fragile to rely on another code path being executed if this is going to be initialized to -1. Also it's more consistent with existing getHeaderTipHeight and getHeaderTipTime to be checking this way

Reply:
I mean this:
```diff
ClientModel::ClientModel(interfaces::Node& node, OptionsModel *_optionsModel, QO
     m_thread->start();

     subscribeToCoreSignals();
+    m_cached_num_blocks = m_node.getNumBlocks();
 }

 ClientModel::~ClientModel()
```
It's the constructor so it's always executed?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/17905#discussion_r384801997

Thanks again for the review!

> I mean this:

Yes, I did understand the suggestion. I like having getNumBlocks implemented the same way as getHeaderTipHeight and getHeaderTipTime, and I like `m_cached_num_blocks` being touched as few places as possible (initialized to -1 and updated only as needed) and not needing to interact with the more complicated constructor.

I also think this case is similar to #18160 where I am trying to avoid an unnecessary interface method call (in that case `interfaces::Wallet::tryGetBalances`, in this case `interfaces::Node::getNumBlocks`), because I see these calls as potentially costly (with #10102 requiring serializing arguments, sending them across a socket and optionally logging them to debug.log, waiting for reply on the socket, deserializing and optionally logging that), but I think you see them more like ordinary method calls.

If other reviewers agree with your suggestion, though, I'm ok with changing this. I don't think it is very significant.

Meta point: When I leave comment on PRs I try be be explicit about (1) What my suggestion is (2) What the benefits of the suggestion are or underlying problem is.

Some reviewers are great at saying what they want but not why they want it, others are great at finding things that should be improved without being specific enough about what kind of change they want to see. In this case, I understand the suggestion but don't understand what's driving it. Readability, avoiding a potential bug, efficiency, or just style?

Reply:
Sorry for not explaining my POV. I'm suggesting to just make `ClientModel::getNumBlocks` lock free - not sure what would be the first call stack hitting the lock and if it would hurt the event loop, but with my suggestion that wouldn't be a concern. But like you say, it's not very significant so feel free to ignore it.